File: carabiner/application.py
import gi
gi.require_version("Gtk", "4.0")
gi.require_version("Adw", "1")
from gi.repository import Gtk, Adw, Gio, GLib

from carabiner.window import CarabinerWindow
from carabiner.backend.events import set_main_thread_dispatcher
from carabiner.backend.settings import load_settings
from carabiner.backend.startup import start_configured_items
from carabiner.backend.tunnel_store import load_tunnels

def gtk_main_thread_dispatcher(callback, *args, **kwargs):
    def wrapper():
        try:
            callback(*args, **kwargs)
        except Exception as e:
            print(f"Exception in UI thread: {e}")
        return GLib.SOURCE_REMOVE
    GLib.idle_add(wrapper)

class CarabinerApplication(Adw.Application):
    def __init__(self):
        super().__init__(
            application_id="io.github.sugarycandybar.Carabiner",
            flags=Gio.ApplicationFlags.HANDLES_COMMAND_LINE
        )
        self._activate_in_background = False
        self._startup_items_started = False
        self._background_hold = False
        set_main_thread_dispatcher(gtk_main_thread_dispatcher)

    def do_command_line(self, command_line):
        args = command_line.get_arguments()
        self._activate_in_background = "--background" in args
        self.activate()
        return 0

    def do_activate(self):
        if self._activate_in_background:
            self._activate_in_background = False
            if not self._background_hold:
                self.hold()
                self._background_hold = True
            self._start_startup_items_once()
            return

        win = self.props.active_window
        if not win:
            win = CarabinerWindow(application=self)
        win.present()
        self._start_startup_items_once()

    def _start_startup_items_once(self):
        if self._startup_items_started:
            return
        self._startup_items_started = True
        settings = load_settings()
        if settings.get("playit_agent_autostart") or any(
            t.get("autostart") and str(t.get("provider", "")).lower() != "playit"
            for t in load_tunnels()
        ):
            start_configured_items()

    def release_background_hold(self):
        if self._background_hold:
            self.release()
            self._background_hold = False

File: src/application.rs
use crate::{
    constants::APP_ID, settings::load_settings, startup::start_configured_items,
    tunnel_store::load_tunnels, window::CarabinerWindow,
};
use adw::prelude::*;
use gtk::{gio, glib};
use std::{
    cell::{Cell, RefCell},
    rc::Rc,
};

#[derive(Default)]
struct AppState {
    activate_in_background: Cell<bool>,
    startup_items_started: Cell<bool>,
    background_hold: RefCell<Option<gio::ApplicationHoldGuard>>,
}

pub fn run() -> glib::ExitCode {
    let app = adw::Application::builder()
        .application_id(APP_ID)
        .flags(gio::ApplicationFlags::HANDLES_COMMAND_LINE)
        .build();

    app.set_accels_for_action("win.preferences", &["<Control>comma"]);
    app.set_accels_for_action("win.show-shortcuts", &["<Control>question"]);
    app.set_accels_for_action("win.quit", &["<Control>q", "<Control>w"]);
    app.set_accels_for_action("win.menu", &["F10"]);
    app.set_accels_for_action("win.add-tunnel", &["<Control>n"]);

    let state = Rc::new(AppState::default());

    {
        let state = state.clone();
        app.connect_command_line(move |app, command_line| {
            let background = command_line
                .arguments()
                .iter()
                .any(|arg| arg.to_string_lossy() == "--background");
            state.activate_in_background.set(background);
            app.activate();
            glib::ExitCode::SUCCESS
        });
    }

    {
        let state = state.clone();
        app.connect_activate(move |app| {
            if state.activate_in_background.get() {
                state.activate_in_background.set(false);
                if state.background_hold.borrow().is_none() {
                    *state.background_hold.borrow_mut() = Some(app.hold());
                }
                start_startup_items_once(&state);
                return;
            }

            if let Some(window) = app.active_window() {
                window.present();
            } else {
                let weak_app = app.downgrade();
                let release_state = state.clone();
                let release_background_hold = Rc::new(move || {
                    let _ = weak_app.upgrade();
                    release_state.background_hold.borrow_mut().take();
                });
                let window = CarabinerWindow::new(app, release_background_hold);
                window.present();
            }
            start_startup_items_once(&state);
        });
    }

    app.run()
}

fn start_startup_items_once(state: &Rc<AppState>) {
    if state.startup_items_started.get() {
        return;
    }
    state.startup_items_started.set(true);

    let settings = load_settings();
    let has_autostart_tunnel = load_tunnels()
        .into_iter()
        .any(|tunnel| tunnel.autostart && tunnel.provider.to_lowercase() != "playit");

    if settings.get_bool("playit_agent_autostart") || has_autostart_tunnel {
        start_configured_items::<fn(usize)>(None);
    }
}
